identity_credential/src/sd_jwt_vc/token.rs

```diff
@@ -0,0 +1,477 @@
+    jwk: &Jwk,
+    hasher: &dyn Hasher,
+    options: &KeyBindingJWTValidationOptions,
+  ) -> Result<()> {
+    self.verify_key_binding(jws_verifier, jwk)?;
+
+    if let Some(requirement) = self.required_key_bind() {
+      if self.key_binding_jwt().is_none() {
+        return Err(Error::Validation(anyhow!(
+          "a key binding was required but none was provided"
+        )));
+      }
+      match requirement {
+        RequiredKeyBinding::Jwk(json_jwk) => {
+          if jwk.to_json_value().unwrap().as_object().unwrap() != json_jwk {
+            return Err(Error::Validation(anyhow!(
+              "key used for signing KB-JWT does not match the key required in this SD-JWT"
+            )));
+          }
+        }
+        RequiredKeyBinding::Kid(kid) | RequiredKeyBinding::Jwu { kid, .. } => jwk
+          .kid()
+          .filter(|id| id == kid)
+          .ok_or_else(|| {
+            Error::Validation(anyhow::anyhow!(
+              "the provided JWK doesn't have required `kid` \"{kid}\""
+            ))
+          })
+          .map(|_| ())?,
+        _ => (),
+      }
+    }
+
+    let Some(kb_jwt) = self.key_binding_jwt() else {
+      return Ok(());
+    };
+    let KeyBindingJWTValidationOptions {
+      nonce,
+      aud,
+      earliest_issuance_date,
+      latest_issuance_date,
+      ..
+    } = options;
+
+    let issuance_date =
+      Timestamp::from_unix(kb_jwt.claims().iat).map_err(|_| Error::Validation(anyhow!("invalid `iat` value")))?;
+
+    if let Some(earliest_issuance_date) = earliest_issuance_date {
+      if issuance_date < *earliest_issuance_date {
+        return Err(Error::Validation(anyhow!(
+          "this KB-JWT has been created earlier than `earliest_issuance_date`"
+        )));
+      }
+    }
+
+    if let Some(latest_issuance_date) = latest_issuance_date {
+      if issuance_date > *latest_issuance_date {
+        return Err(Error::Validation(anyhow!(
+          "this KB-JWT has been created later than `latest_issuance_date`"
+        )));
+      }
+    } else if issuance_date > Timestamp::now_utc() {
+      return Err(Error::Validation(anyhow!("this KB-JWT has been created in the future")));
+    }
+
+    if let Some(nonce) = nonce {
+      if nonce != &kb_jwt.claims().nonce {
+        return Err(Error::Validation(anyhow!("invalid KB-JWT's nonce: expected {nonce}")));
+      }
+    }
+
+    if let Some(aud) = aud {
+      if aud != &kb_jwt.claims().aud {
+        return Err(Error::Validation(anyhow!("invalid KB-JWT's `aud`: expected \"{aud}\"")));
+      }
+    }
+
+    // Validate SD-JWT digest.
+    if self.claims()._sd_alg.as_deref().unwrap_or(SHA_ALG_NAME) != hasher.alg_name() {
+      return Err(Error::Validation(anyhow!("invalid hasher")));
+    }
+    let encoded_sd_jwt = self.to_string();
+    let digest = {
+      let last_tilde_idx = encoded_sd_jwt.chars().rev().find_position(|ch| *ch == '~').unwrap().0;
+      let sd_jwt_no_kb = &encoded_sd_jwt[..=last_tilde_idx];
```

Comment:
Just noticed that that is likely not doing what it is intended to do. Since `encoded_sd_jwt` is being reversed, `last_tilde_idx` is the position of the first occurrence of `~` counted from the end of the String. 

Then that position is used to create a slice from the beginning of `encoded_sd_jwt` until `last_tilde_idx`. 

Basically this is what it is doing:
```rust
let encoded_sd_jwt = "example~123";
let last_tilde_idx = encoded_sd_jwt.chars().rev().find_position(|ch| *ch == '~').unwrap().0;
assert_eq!(last_tilde_idx, 3);
let sd_jwt_no_kb = &encoded_sd_jwt[..=last_tilde_idx];
assert_eq!(sd_jwt_no_kb, "exam");
```

This will get you the actual position of the last occurrence of the `~` character:
```rust
let last_tilde_idx = encoded_sd_jwt.rfind('~').unwrap();
```

Reply:
@nanderstabel fixed. Thank you for taking the time